R/as_model_out_df.R

```diff
@@ -0,0 +1,211 @@
+#' already exist in `tbl`, the `tbl` column name containing `output_type` data.
+#' @param output_type_id_col character string. If an `output_type_id` column does not
+#' already exist in `tbl`, the `tbl` column name containing `output_type_id` data.
+#' @param value_col character string. If an `value` column does not
+#' already exist in `tbl`, the `tbl` column name containing `value` data.
+#' @param sep character string. Character used as separator when concatenating
+#' `team_abbr` and `model_abbr` column values into a single `model_id` string. Only
+#' applicable if `model_id` column not present and `team_abbr` and `model_abbr` columns are.
+#' @param trim_to_task_ids logical. Whether to trim `tbl` to task ID columns only. Task ID
+#' columns can be specified by providing a `<hub_connection>` class object to `hub_con` or
+#' manually through `task_id_cols`.
+#' @param hub_con a `<hub_connection>` class object. Only used if
+#' `trim_to_task_ids = TRUE` and tasks IDs should be determined from the hub config.
+#' @param task_id_cols a character vector of column names. Only used if
+#' `trim_to_task_ids = TRUE` to manually specify task ID columns to retain.
+#' Overrides `hub_con` argument if provided.
+#' @param remove_empty Logical. Whether to remove columns containing only `NA`.
+#'
+#' @return A `model_out_tbl` class object.
+#' @export
+#'
+#' @examples
+#' library(dplyr)
+#' hub_path <- system.file("testhubs/flusight", package = "hubUtils")
+#' hub_con <- connect_hub(hub_path)
+#' hub_con %>%
+#'   filter(output_type == "quantile", location == "US") %>%
+#'   collect() %>%
+#'   filter(forecast_date == max(forecast_date)) %>%
+#'   as_model_out_tbl()
+#'
+as_model_out_tbl <- function(tbl, model_id_col = NULL, output_type_col = NULL,
+                             output_type_id_col = NULL, value_col = NULL, sep = "-",
+                             trim_to_task_ids = FALSE, hub_con = NULL,
+                             task_id_cols = NULL, remove_empty = FALSE) {
+  checkmate::assert_data_frame(tbl)
+  tbl <- tibble::as_tibble(tbl)
+
+  tbl <- rename_columns(tbl,
+    model_id_col = model_id_col,
+    output_type_col = output_type_col,
+    output_type_id_col = output_type_id_col,
+    value_col = value_col
+  )
+
+  if (!"model_id" %in% names(tbl)) {
+    cli::cli_alert_warning("{.arg model_id} column missing. Attempting to create automatically.")
+    tbl <- model_id_merge(tbl, sep = sep)
+  }
+
+  if (trim_to_task_ids) {
+    tbl <- trim_tbl_to_task_ids(
+      tbl = tbl,
+      task_id_cols = task_id_cols,
+      hub_con = hub_con
+    )
+  }
+
+  if (remove_empty) {
+    tbl <- remove_model_out_tbl_empty(tbl)
+  }
+
+  tbl <- std_col_order_model_out_tbl(tbl) %>%
+    std_row_order_model_out_tbl()
+
+  validate_model_out_tbl(tbl)
+
+  structure(tbl,
+    class = c("model_out_tbl", class(tbl))
+  )
+}
+
+# --- Unexported Utilities ----
+validate_model_out_tbl <- function(tbl) {
+  if (!all(std_colnames %in% names(tbl))) {
```

Comment:
this might be excessive -- but a slightly more precise check here might be something like "each of the `std_colnames` appears in `names(tbl)` exactly once"

Reply:
By default, both `data.frame`s and `tibble`s must have unique column names so the `exactly once` aspect of the check is guaranteed through the class itself.